cmd/client/httpproxy/httpproxy.go

```
@@ -0,0 +1,316 @@
+func (sm *SessionManager) addSession(session *yamux.Session, logStream *yamux.Stream) {
+	sm.sessions = append(sm.sessions, session)
+	sm.loggingStream = append(sm.loggingStream, logStream)
+	go func() {
+		<-session.CloseChan() // Wait for the session to close
+		sm.removeSession(session, logStream)
+	}()
+}
+
+// removeSession is a simple function that removes a session from the slice
+func (sm *SessionManager) removeSession(session *yamux.Session, logStream *yamux.Stream) {
+	for i, s := range sm.sessions {
+		if s == session {
+			log.Printf("one of the upstream connections went down and has been removed. Pool size is now %d\n", len(sm.sessions))
+			sm.sessions = append(sm.sessions[:i], sm.sessions[i+1:]...)
+			sm.loggingStream = append(sm.loggingStream[:i], sm.loggingStream[i+1:]...) // Remove the corresponding logging stream
+
+			// Todo, we could try to reconnect here and add one back to the pool
+			break
+		}
+	}
+	if len(sm.sessions) == 0 {
+		log.Fatalf("All upstream sessions have been closed, no more upstream options available. Exiting.")
+	}
+}
+
+// selectSession is a simple function that selects a session from the slice
+func (sm *SessionManager) selectSession(index int) *yamux.Session {
+	return sm.sessions[index%len(sm.sessions)]
+}
+
+// handleClientConnection is a function that handles the client connection
+// func handleClientConnection(clientConn net.Conn, session *yamux.Session) {
+func handleClientConnection(clientConn net.Conn, session *yamux.Session, updateTextView func(string)) {
+
+	defer clientConn.Close()
+	//fmt.Println("New connection accepted from ", clientConn.RemoteAddr())
+
+	//logLine := fmt.Sprintf("New connection accepted from %s", clientConn.RemoteAddr())
+	//updateTextView(logLine)
+
+	stream, err := session.OpenStream()
+	if err != nil {
+		log.Printf("Failed to open yamux stream: %v", err)
+		return
+	}
+	defer stream.Close()
+
+	go io.Copy(clientConn, stream)
```

ignored error conditions here.